### PR TITLE
Trajectory balance loss and other improvements

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**
The core thing of this PR is the new trajectory balance loss for GFlowNet, but there are multiple other small improvements here. I apologies for the lack of the documentation of them here.

### CODE REVIEW

File: gflownet.py [hunk 1]
```diff
@@ -205,11 +218,22 @@ def set_device(dev):
 class GFlowNetAgent:
     def __init__(self, args, comet=None, proxy=None, al_iter=-1, data_path=None):
         # Misc
-        self.rng = np.random.RandomState(int(time.time()))
+        self.rng = np.random.default_rng(int(time.time()))
```
Comment (InfluenceFunctional) [inline gflownet.py:221]: should this be settable e.g. by the global sampler seed?

File: gflownet.py [hunk 2]
```diff
@@ -137,6 +148,8 @@ def add_args(parser):
     args2config.update({"clip_grad_norm": ["gflownet", "clip_grad_norm"]})
     parser.add_argument("--random_action_prob", default=0.0, type=float)
     args2config.update({"random_action_prob": ["gflownet", "random_action_prob"]})
+    parser.add_argument("--pct_batch_empirical", default=0.0, type=float)
```
Comment (InfluenceFunctional) [inline gflownet.py:151]: not sure how to use this if nonzero
had some trouble in the past getting GFN to run because of missing args to do with new GFN test / eval metrics

File: gflownet.py [hunk 3]
```diff
@@ -205,11 +218,22 @@ def set_device(dev):
...
         self.debug = args.debug
         self.device_torch = torch.device(args.gflownet.device)
         self.device = self.device_torch
         set_device(self.device_torch)
+        if args.gflownet.loss in ["flowmatch"]:
+            self.loss = "flowmatch"
+            self.Z = None
+        elif args.gflownet.loss in ["trajectorybalance", "tb"]:
+            self.loss = "trajectorybalance"
+            self.Z = nn.Parameter(torch.ones(64) * 150.0 / 64)
```
Comment (InfluenceFunctional) [inline gflownet.py:231]: reason for choice? I didn't see / might have missed in the paper

File: gflownet.py [hunk 4]
```diff
@@ -544,6 +584,45 @@ def learn_from(self, it, batch):
...
+        ]
+        # Sum of log probs
+        sumlogprobs = tf(
+            torch.zeros(len(torch.unique(traj_id, sorted=True)))
+        ).index_add_(0, traj_id, logprobs)
+        # Sort rewards of done sequences by ascending traj id
+        rewards = rewards[done.eq(1)][torch.argsort(traj_id[done.eq(1)])]
+        # Trajectory balance loss
+        loss = (self.Z.sum() + sumlogprobs - torch.log((rewards))).pow(2).mean()
+        return loss, loss, loss
```
Comment (InfluenceFunctional) [inline gflownet.py:624]: probably need to work through this myself to understand the implementation completely, but it looks ok

**Comment by InfluenceFunctional** [comment]
Overall looks sensible. Couple of open questions around setting up appropriate empirical losses (not necessary with trajectory loss?) and test metrics in the default yaml config(s).

If performance is as advertised in the paper, very excited to play with trajectory loss. Seems really really cool.

File: gflownet.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply gflownet.py:221]: Yes, indeed. I have opened an issue [here](https://github.com/InfluenceFunctional/ActiveLearningPipeline/issues/44#issue-1143864976) to keep track.

File: gflownet.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply gflownet.py:151]: You mean how to actually set a different value than the default? Two ways, as with every other parameter unless badly implemented:
* Via the command line: `--pct_batch_empirical 0.2`
* Via the config file: add `pct_batch_empirical: 0.2` in the `gflownet` part of the config file (see [here](https://github.com/InfluenceFunctional/ActiveLearningPipeline/pull/43/files#diff-0aea26b2da009a99c51157b577f186bb59b5ad6770e10e2945c61acb98a85e9fR93), for instance):
``` yaml
...
gflownet:
...
  pct_batch_empirical: 0.2
```

File: gflownet.py (hunk 3 above)
Comment (alexhernandezgarcia) [reply gflownet.py:231]: You mean the specific values right? Heuristic recommended by Moksh, no other reason...

File: gflownet.py (hunk 4 above)
Comment (alexhernandezgarcia) [reply gflownet.py:624]: Yeah, the way the variables are computed is not super intuitive because efficiency is favoured. I hope it is correct, sanity checks always welcome :)

File: gflownet.py (hunk 3 above)
Comment (InfluenceFunctional) [reply gflownet.py:231]: okidoke
